src/vivarium/framework/components/manager.py

```diff
@@ -26,6 +26,60 @@ class ComponentConfigError(VivariumError):
     pass
 
 
+class OrderedComponentSet:
+    """A container for Vivarium components. It preserves ordering, enforces
+    uniqueness by name, and provides a subset of set-like semantics."""
+
+    def __init__(self, *args):
+        self.names = set()
+        self.components = []
+        if args:
+            self.update(args)
+
+    def add(self, component: Any) -> None:
+        if component in self:
+            raise ComponentConfigError(f"Attempting to add a component with duplicate name: {component}")
+        self.names.add(component.name)
+        self.components.append(component)
+
+    def update(self, components: Union[List, Tuple]):
+        for c in components:
+            self.add(c)
+
+    def __contains__(self, component: Any) -> bool:
+        if not hasattr(component, "name"):
+            raise ComponentConfigError(f"Component {component} has no name attribute")
+        return component.name in self.names
+
+    def __iter__(self) -> Iterator:
+        return iter(self.components)
+
+    def __len__(self) -> int:
+        return len(self.components)
+
+    def __bool__(self) -> bool:
+        return bool(self.components)
+
+    def __eq__(self, other) -> bool:
+        try:
```

Comment:
I suggest doing:

```
try:
    return type(self) is type(other) and self.names == other.names
except TypeError:
    return False

Reply:
but comparing names won't check order and that seems like an important condition for equality here?

Reply:
Since ordering matters I'm going to convert names to a list and then do this

Reply:
just want to make sure I'm understanding things here: I understood why we had names as a set to make the lookups fast but is the reason we still keep the names separate as a list (even though they're also contained in the components list) to make it so we don't have to do explicit list comprehensions for contains/equality and can just pass it on to the list methods?

Reply:
it's just kruft now